crates/stackable-certs/src/ca/ca_builder.rs

```diff
@@ -0,0 +1,240 @@
+        //
+        // The root profile doesn't add the AuthorityKeyIdentifier extension.
+        // We manually add it below by using the 160-bit SHA-1 hash of the
+        // subject public key. This conforms to one of the outlined methods for
+        // generating key identifiers outlined in RFC 5280, section 4.2.1.2.
+        //
+        // Prepare extensions so we can avoid clones.
+        let aki = AuthorityKeyIdentifier::try_from(spki.owned_to_ref())
+            .context(ParseAuthorityKeyIdentifierSnafu)?;
+
+        let signer = signing_key_pair.signing_key();
+        let mut builder = CertificateBuilder::new(
+            Profile::Root,
+            serial_number,
+            validity,
+            subject,
+            spki,
+            signer,
+        )
+        .context(CreateCertificateBuilderSnafu)?;
+
+        builder
+            .add_extension(&aki)
+            .context(AddCertificateExtensionSnafu)?;
+
+        debug!("creating and signing CA certificate");
+        let certificate = builder.build().context(BuildCertificateSnafu)?;
+
+        Ok(CertificateAuthority {
+            certificate_pair: CertificatePair {
+                certificate,
+                key_pair: signing_key_pair,
+            },
+        })
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use x509_cert::certificate::TbsCertificateInner;
+
+    use super::*;
+    use crate::keys::{ecdsa, rsa};
+
+    #[test]
+    fn minimal_ca() {
+        let ca: CertificateAuthority<ecdsa::SigningKey> = CertificateAuthorityBuilder::builder()
+            .build()
+            .build_ca()
+            .expect("failed to build CA");
+
+        assert_ca_cert_attributes(
+            &ca.ca_cert().tbs_certificate,
+            SDP_ROOT_CA_SUBJECT,
+            DEFAULT_CA_VALIDITY,
+            None,
+        )
+    }
+
+    #[test]
+    fn customized_ca() {
+        let ca = CertificateAuthorityBuilder::builder()
+            .subject("CN=Test")
+            .serial_number(42)
+            .signing_key_pair(rsa::SigningKey::new().unwrap())
+            .validity(Duration::from_days_unchecked(13))
+            .build()
+            .build_ca()
+            .expect("failed to build CA");
+
+        assert_ca_cert_attributes(
+            &ca.ca_cert().tbs_certificate,
+            "CN=Test",
+            Duration::from_days_unchecked(13),
+            Some(42),
+        )
+    }
+
+    fn assert_ca_cert_attributes(
+        ca_cert: &TbsCertificateInner,
+        subject: &str,
+        validity: Duration,
+        serial_number: Option<u64>,
+    ) {
+        assert_eq!(ca_cert.subject, subject.parse().unwrap());
+
+        let not_before = ca_cert.validity.not_before.to_system_time();
+        let not_after = ca_cert.validity.not_after.to_system_time();
+        assert_eq!(
+            not_after
+                .duration_since(not_before)
+                .expect("Failed to calculate duration between notBefore and notAfter"),
```

Comment:
praise: Thanks for using expect to explain why it should be valid in the test :+1: